### ISSUE TITLE
Lag.

### ISSUE THREAD

**Issue by NatsuTheGreat** [Barbosik/MultiOgar#115]
Ejected mass seems to really cause lag. There were about 1000 ejected mass cells and three players. Update time: 55+ MS. 130 Viruses, and three players = no lag. Can this be optimized?

**Comment by Barbosik** [comment]
yes, but there is need to rewrite collision detection and entity management

**Comment by MastaCoder** [comment]
@Barbosik Yep it has to do with the collisions is what I can confirm. I removed it from my site and the lag was completely gone. (on a different server software.) Might need a lot of rewriting :p

**Comment by NatsuTheGreat** [comment]
It could ```updateMoveEngine``` or rigid collisions.

**Comment by Barbosik** [comment]
updateMoveEngine and nodesEjected

**Comment by Barbosik** [comment]
you can change loop count from 2 to 1 on rigid body collision resolution, it will improve perfromance, but cells jitter may be worse (when there are a lot of small cells).

**Comment by NatsuTheGreat** [comment]
Sorry, what function?

**Comment by NatsuTheGreat** [comment]
http://prntscr.com/bog4sh This is also an issue.

**Comment by NatsuTheGreat** [comment]
http://prntscr.com/bog4xm

**Comment by Barbosik** [comment]
I think it's not important for ejected cells. But with no external force they should move in such way to not intersect with each other. If they stops to move and still intersect each other - it's a bug

**Comment by NatsuTheGreat** [comment]
Jittering is caused by `null` of collision and extrusion force. Pretty sure. And I'll see what I can do about the ejected mass problem.

**Comment by Barbosik** [comment]
Jittering caused because rigid body collision resolved with no take into account velocity of the body.
So, when several rigid bodies affected with force (from mouse) they overlap each other, because we didn't corrected velocity vector before applying to the body position

**Comment by Barbosik** [comment]
There is need to calculate velocity first, then check if a new rigid body bounds will collide with other rigid bodies and apply resolution impulse to the velocity instead of position. When all bodies will be processed, we can apply corrected velocity to the body position. It should works in such way :)

**Comment by NatsuTheGreat** [comment]
I actually removed an if statement, it seems that most if the jittering is fixed :D But cell squeezing is there, sometimes makes small squeeze and move around rapidly. (the little cells are surrounded by BIG cells)

**Comment by Barbosik** [comment]
if you remove statement `if (manifold==null)` it will works incorrectly, because it will process collisions for objects with square shape instead of circle :)
But there is a thing related with collision distance threshold. Currently there is no threshold, but physics engines using it to eliminate jittering. So, there is possible experiment with checkCellCollision - you can try to add some threshold to not trigger collision if distance is too small. I'm not sure if it will fix jittering, but may be it can help

**Comment by nobsterr** [comment]
not laggy , you can test it on agario.city , Multiogar more 100 users

**Comment by NatsuTheGreat** [comment]
I didn't removed that if statement. I removed ```if (squared > r * r) return null;``` Jittering seems to be well gone.
